### PR TITLE
Further JSON abstraction

### LINKED ISSUE #90: Encapsulate the configuration and command JSON parsing
The configuration and command related JSON parsing would be better if placed in dedicated classes.
This would decrease the coupling in the code and make it easier to write unit tests.

### PR DESCRIPTION
**Pull request by mattclarke**
### Description of work

The code for parsing the configuration files and handling commands has been extracted into new classes to enable a separation of concerns.

### Issue
Closes #90

### Acceptance Criteria
The actual functional code hasn't change very much as it is mostly been extracted into separate classes unchanged. Check that using a configuration file and sending commands still works as before.

### Unit Tests
The unit tests have been updated to use the new classes and the source files renamed.

### Other

A system test has been added which checks that a channel defined in the configuration file is enabled correctly.

The README has been updated to give examples of all the command types.

---

## Code Review (To be filled in by the reviewer only)

- [x] Is the code of an acceptable quality?
- [x] Do the changes function as described and is it robust?

---

## Nominate for Group Code Review (Anyone can nominate it)
Indicate if you think the code should be reviewed in a Thursday code review session.

- [x] Recommend for group code review

Also, nominate it on the code_review Slack channel (does someone want to automate this?).

### CODE REVIEW

File: src/ConfigParser.cpp [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+}
+
+void ConfigParser::extractMappingInfo(nlohmann::json const &Mapping,
+                                      std::string &Channel,
+                                      std::string &Protocol) {
+  if (!Mapping.is_object()) {
+    throw MappingAddException("Given Mapping is not a JSON object");
+  }
+
+  auto ChannelMaybe = find<std::string>("channel", Mapping);
```
Comment (rerpha) [inline src/ConfigParser.cpp]: wrap in 

`if (auto ChannelMaybe = find<std::string>("channel", Mapping)) {
 Channel = ChannelMaybe.inner}
 else throw MappingAddException("Cannot find channel")`

Also, I think we should rename `ChannelMaybe` to `FoundChannel` or something. 

typo in "can not find channel" - should be "cannot find channel"

File: src/ConfigParser.cpp [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+  } else {
+    // Default is pva
+    Protocol = "pva";
+  }
+}
+
+ConverterSettings
+ConfigParser::extractConverterSettings(nlohmann::json const &Mapping) {
+  ConverterSettings Settings;
+  Settings.Schema = find<std::string>("schema", Mapping).inner();
```
Comment (rerpha) [inline src/ConfigParser.cpp]: should really have error handling like the rest of the `find` checks.

File: system-tests/docker-compose.yml [hunk 3]
```diff
@@ -37,6 +38,7 @@ services:
       - kafka
       - ioc
     network_mode: "host"
+    volumes:
+      - ./forwarder_config.json:/forwarder_config.json
     environment:
-      CONFIG_URI: "//localhost:9092/TEST_forwarderConfig"
```
[the remark this answers is not in the document]
Comment (mattclarke) [reply system-tests/docker-compose.yml]: These are now defined in the config file. Still have the same values

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -85,7 +86,8 @@ schemas/f142/f142.cpp
 ${FMT_SRC}
 schemas/f143/f143.cpp
 ${CMAKE_CURRENT_BINARY_DIR}/git_commit_current.cpp
-)
+CommandHandler.h
```
[the remark this answers is not in the document]
Comment (mattclarke) [reply src/CMakeLists.txt]: Better to wait until you merge your work.

File: src/ConfigParser.cpp (hunk 1 above)
Comment (mattclarke) [reply src/ConfigParser.cpp]: Good call!

File: src/ConfigParser.cpp (hunk 2 above)
Comment (mattclarke) [reply src/ConfigParser.cpp]: 👍

File: src/CommandHandler.cpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+#include "ConfigParser.h"
+#include "helper.h"
+#include "json.h"
+#include "logger.h"
+#include <nlohmann/json.hpp>
+
+namespace BrightnESS {
+namespace ForwardEpicsToKafka {
+
+ConfigCB::ConfigCB(Main &main) : main(main) {}
```
Comment (SkyToGround) [inline src/CommandHandler.cpp]: More descriptive names? There are a lot of mains!

File: src/ConfigParser.h [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  /// Extract the configuration information from the JSON.
+  ///
+  /// \return The extracted settings.
+  ConfigSettings extractConfiguration();
+
+  /// Set the brokers in the supplied settings.
+  ///
+  /// \param Brokers The raw brokers information, e.g. "localhost:9092"
+  /// \param Settings The settings to write the brokers to.
+  static void setBrokers(std::string const &Brokers, ConfigSettings &Settings);
```
Comment (SkyToGround) [inline src/ConfigParser.h]: Add documentation on which brokers are set.

File: src/tests/ConfigParser_tests.cpp [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+  Config.setJsonFromString(RawJson);
+
+  BrightnESS::ForwardEpicsToKafka::ConfigSettings Settings =
+      Config.extractConfiguration();
+
+  ASSERT_EQ(1u, Settings.GlobalConverters.size());
+
+  auto &f142 = Settings.GlobalConverters["f142"];
+  ASSERT_EQ(2u, f142.ConfigurationIntegers.size());
+  ASSERT_EQ(123, f142.ConfigurationIntegers["some_option1"]);
```
Comment (dominikwerder) [inline src/tests/ConfigParser_tests.cpp]: Harden the test using `.at()` ?

File: src/tests/ConfigParser_tests.cpp [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+                        "  \"broker\": \"kafkabroker:1234\""
+                        "}";
+
+  BrightnESS::ForwardEpicsToKafka::ConfigParser Config;
+  Config.setJsonFromString(RawJson);
+
+  BrightnESS::ForwardEpicsToKafka::ConfigSettings Settings =
+      Config.extractConfiguration();
+
+  ASSERT_EQ("kafkabroker", Settings.Brokers[0].host);
```
Comment (dominikwerder) [inline src/tests/ConfigParser_tests.cpp]: Avoid UB by `.at(0)` ?

**Pull request merged**